src/Microsoft.AspNetCore.SignalR.Core/HubConnectionContext.cs

```diff
@@ -86,33 +88,56 @@ public class HubConnectionContext
 
         public int? LocalPort => Features.Get<IHttpConnectionFeature>()?.LocalPort;
 
-        public async Task WriteAsync(HubMessage message, bool throwOnFailure = false)
+        public virtual async Task WriteAsync(HubMessage message)
         {
-            while (await _output.Writer.WaitToWriteAsync())
+            var buffer = ProtocolReaderWriter.WriteMessage(message);
+
+            // Hubs support multiple producers so need to lock as we write into the buffer
+            // since the pipe only supports a single consumer at a time
+            lock (_writeLock)
             {
-                if (_output.Writer.TryWrite(message))
-                {
-                    return;
-                }
+                _connectionContext.Transport.Output.Write(buffer);
             }
 
-            _logger.OutboundChannelClosed();
+            Interlocked.Exchange(ref _lastSendTimestamp, Stopwatch.GetTimestamp());
 
-            if (throwOnFailure)
+            await FlushAsync(_connectionContext.Transport.Output, CancellationToken.None);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private Task FlushAsync(PipeWriter pipeWriter, CancellationToken cancellationToken)
+        {
+            var awaitable = pipeWriter.FlushAsync(cancellationToken);
+            if (awaitable.IsCompleted)
             {
-                throw new OperationCanceledException("Outbound channel was closed while trying to write hub message");
+                // The flush task can't fail today
```

Comment:
micro-nit: Don't put temporal references in comments, it'll just get out of date ;P.